### ISSUE TITLE
Suggestion: `no-restricted-globals` for types

### ISSUE THREAD

**Issue by OliverJAsh** [typescript-eslint/typescript-eslint#673]
The ESLint rule [`no-restricted-globals`](https://eslint.org/docs/rules/no-restricted-globals) is great, but it only applies to global values. I want the same thing, but for global _types_.

Example usage: I want to [disallow usage of the global TypeScript `Omit` type because it's not strict enough](https://github.com/microsoft/TypeScript/issues/30825#issuecomment-508584855), but I still want to allow imported types of that same name to be used.

``` ts
// error
type F = Omit<{}, 'foo'>
```

``` ts
import { Omit } from './helpers';

// no error
type F = Omit<{}, 'foo'>
```

This is similar to the existing [`ban-types`](https://github.com/typescript-eslint/typescript-eslint/blob/af70a5932f8032cde13f4a0c98d6657cdb40b441/packages/eslint-plugin/docs/rules/ban-types.md) rule, but different because `ban-types` bans the type both globally _and_ locally.

**Comment by bradzacher** [comment]
I would think a better solution is to rename your helpers so that they don't clash, so that there is no weirdness, and a rule that catches non-imported vs imported isn't needed.

Having developers get confused that they're using "the wrong Omit" is a pretty clear smell IMO:
- your editor can't suggest that it be auto-imported, because it's already a thing in the global namespace
- the compiler won't catch it for the same reason
- you can't auto-fix it because eslint rules shouldn't be adding imports as well as changing other code

**Comment by OliverJAsh** [comment]
I agree, but sometimes we don't get to decide how to name things—libraries do it for us.

Examples:

- the popular [`history`](https://github.com/ReactTraining/history) module has a `Location` type, but there's also a global `Location`
- many libraries ([fp-ts](https://github.com/gcanti/fp-ts), [Funfix](https://github.com/funfix/funfix)) provide an `Option` type, but there's also a global `Option` (modelling something entirely different)

In that case, we still want to catch mistakes where the global one is being used when it shouldn't.

**Comment by OliverJAsh** [comment]
> * your editor can't suggest that it be auto-imported, because it's already a thing in the global namespace

Side note: I really want this VS Code feature to help in this case: https://github.com/microsoft/TypeScript/issues/31199.

**Comment by vladimiry** [comment]
The type provided by the library can be renamed and exported to global scope, [one example](https://github.com/vladimiry/ElectronMail/blob/375f760d5d801da65b8b2d41adcfef0892a2e5eb/src/%40types/index.d.ts#L18).

**Comment by OliverJAsh** [comment]
Hmm, that's only reasonable if you're happy polluting the global namespace. Seems a shame to me to have to sacrifice modules/imports.

Also, sometimes you're dealing with not just a type but also a value, e.g. the `Option` I gave is a class.

There are different ways of managing this problem. This lint rule suggestion would enable one of them, which some people might prefer.

**Comment by OliverJAsh** [comment]
Regardless of how you avoid the name conflict, we still have the core problem of "how to prevent accidental usage of the global type". Even if I name my custom `Omit` something else like `StrictOmit`, I still want to disallow accidental usage of the global `Omit`. We can use `ban-types` for that, but that's overly specific—we only need to ban the global type.

**Comment by OliverJAsh** [comment]
All of the reasons you might want this for values also apply to types. So if ESLint can offer it for values, shouldn't this plugin provide parity by offering it for TS types?

**Comment by OliverJAsh** [comment]
Just came across this again. I can ban `location` using `no-restricted-globals` but still use that identifier for non-global values (✅), however if I ban the global type `Location` using `ban-types` it will also ban that identifier for non-global values (❌).

If I wanted to write this lint rule, how can I read a list of "global types" in the lint rule? `no-restricted-globals` uses `scope.through`—is there a type equivalent of this?

https://github.com/eslint/eslint/blob/9dfc8501fb1956c90dc11e6377b4cb38a6bea65d/lib/rules/no-restricted-globals.js#L113

**Comment by bradzacher** [comment]
It depends on what you mean by "global types".

There are two types of global types:
1) typescript's core lib types (provided in `lib.**.d.ts`, specified by the `lib` compiler option)
2) user-land global types:
  a) types defined in `@types/**`
  b) types defined in misc packages in `node_modules`
  c) project-local types.

For (1), it's super simple to detect them, see this example from the `no-throw-literal` rule, which detects if the given type is the `Error` type defined in a TS core lib.

https://github.com/typescript-eslint/typescript-eslint/blob/6aa6bc738705927685e1c0dfa12ccadc3fe53d96/packages/eslint-plugin/src/rules/no-throw-literal.ts#L31-L40

For (2), it gets a bit difficult. For example, people are trying to amend the `unbound-method` rule so it doesn't error on `console.*` methods. But the `console` is a core lib type for dom, but not for nodejs. Related comment:
https://github.com/typescript-eslint/typescript-eslint/issues/1085#issuecomment-571906463

In this case, you have to default to tracing the symbol to figure out where it originated. i.e. if it was defined locally in any way (imported, type def, generic, etc), then it's not a global type.
This also has a bit of weirdness due to [interface/namespace declaration merging](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgDIHsFzMdJkDeAUMqcgO7roBeANhOXNXRAFzIDOYUoA5gNxEAvkSIATCAlpwoKBHi7IAHuwxYceQaKUA6SjXqNm9HaCkBXCRwAUAclsBKfsgD0L5CHTJoUdFAA0yGLmKGBeElLIALbQvHxAA).

**Comment by OliverJAsh** [comment]
Thank you @bradzacher. In my specific case I only care about (1), so that's very useful 😄

**Comment by lizozom** [comment]
Hey @OliverJAsh @bradzacher,

I'm working on implementing `ApiExtractor` in `Kibana`. 
We ran into a complicated issue last week with the build process:

@octogonz kindly [helped us determine](https://github.com/microsoft/rushstack/issues/1765) that the build process failed, because we used a **type** from `React` in a `.ts` file. In that file, `React` [wasn't imported explicitly](https://github.com/elastic/kibana/pull/56955/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx#diff-cf8ef695b521693ffb9a53d9c89d41d2R20), so the fallback was to use the global `React`, causing the `ApiExtractor` failure.

We would like to add a `typescript-eslint` rule to our project, preventing any developer from using `React` types, without importing `React`. Since there is no `no-restricted-globals` for types, do you see any possible workarounds for this situation?

**Comment by bradzacher** [comment]
There's no real solution right now.
It would be relatively trivial to write an extension rule for `no-restricted-globals` which analyses `type` references to check if:
1) the name is imported or not.
2) the name is not "banned", if it's not imported.

There's already example code in the plugin for doing check (1):
https://github.com/typescript-eslint/typescript-eslint/blob/133f622f38a286eac45288a9789d2ee529d5e582/packages/eslint-plugin/src/rules/unbound-method.ts#L97-L111

Happy to accept a PR to add this in!

**Comment by lizozom** [comment]
@bradzacher any examples of **extension rule**s for reference?

**Comment by bradzacher** [comment]
Many!
https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#extension-rules

One of the simple examples is the `no-dupe-class-members` rule

**Comment by runspired** [comment]
Seems like this was added recently since bumping to latest caused typed returns from functions that are returning values gotten from native functions (that are returning native promises) to error.

**Comment by OliverJAsh** [comment]
I think this makes more sense as an option in `ban-types` rather than an extension of `no-restricted-globals`, because `no-restricted-globals` is currently only concerned with values whereas `ban-types` is concerned with types. I have a WIP PR to add an option called `banGlobalOnly` to the `ban-types` rule: https://github.com/typescript-eslint/typescript-eslint/pull/4441

**Comment by bradzacher** [comment]
When you first posted this issue back in 2019, we didn't have our own scope manager.
So types weren't properly handled as "variables" in regards to ESLint.
However - now we do.

So [`no-restricted-globals` now works for this usecase of banning global types](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&showAST=scope&code=PQKgBApgzgNglgOwC5gQewLQCdpK3AYyQgBMMBzGNAIwEMYoAuMAbQHIIss0s2AaMGwCqiJAEYAbAEEutAJ4BhNAih4ArkR79BAcQgJOtJDwBiahETjKlK9Zt4BdMCGABYAFAekcgA4QwACpiYAC8YCLIkjJY8jaqWBrGWADcXr7+AQBMoWB6BjFJZhZIVghxdknJQA&rules=N4XyA&tsConfig=N4XyA)!
We don't need a new rule or new options to have this work!

**Comment by OliverJAsh** [comment]
Amazing! I guess we can close this then?
